File: playground/cpp/days/day03/tests/event_bus_test.cpp
#include "day03/event_bus.h"
#include "run_together.h"
#include <gtest/gtest.h>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <future>
#include <set>
#include <string>
#include <thread>
#include <vector>

using namespace std::chrono_literals;

TEST(EventBus, PublishCallsSubscribersInOrder) {
    EventBus bus;
    std::vector<std::string> log;
    bus.subscribe([&](const std::string& e) { log.push_back("A:" + e); });
    bus.subscribe([&](const std::string& e) { log.push_back("B:" + e); });
    bus.subscribe([&](const std::string& e) { log.push_back("C:" + e); });
    EXPECT_EQ(bus.publish("x"), 3u);
    EXPECT_EQ(log, (std::vector<std::string>{"A:x", "B:x", "C:x"}));
}

TEST(EventBus, PublishWithNoSubscribersCallsNobody) {
    EventBus bus;
    EXPECT_EQ(bus.publish("x"), 0u);
    EXPECT_EQ(bus.subscriberCount(), 0u);
}

TEST(EventBus, IdsArePositiveAndUnique) {
    EventBus bus;
    std::set<long long> ids;
    for (int i = 0; i < 100; ++i) {
        long long id = bus.subscribe([](const std::string&) {});
        EXPECT_GT(id, 0);
        ids.insert(id);
    }
    EXPECT_EQ(ids.size(), 100u);
    EXPECT_EQ(bus.subscriberCount(), 100u);
}

TEST(EventBus, UnsubscribeStopsDelivery) {
    EventBus bus;
    std::vector<std::string> log;
    bus.subscribe([&](const std::string&) { log.push_back("A"); });
    long long b = bus.subscribe([&](const std::string&) { log.push_back("B"); });
    bus.subscribe([&](const std::string&) { log.push_back("C"); });

    EXPECT_TRUE(bus.unsubscribe(b));
    EXPECT_FALSE(bus.unsubscribe(b)) << "already removed";
    EXPECT_FALSE(bus.unsubscribe(12345)) << "never existed";
    EXPECT_EQ(bus.subscriberCount(), 2u);
    EXPECT_EQ(bus.publish("x"), 2u);
    EXPECT_EQ(log, (std::vector<std::string>{"A", "C"}));
}

TEST(EventBus, OrderIsKeptAfterUnsubscribeAndSubscribe) {
    EventBus bus;
    std::vector<std::string> log;
    bus.subscribe([&](const std::string&) { log.push_back("A"); });
    long long b = bus.subscribe([&](const std::string&) { log.push_back("B"); });
    bus.subscribe([&](const std::string&) { log.push_back("C"); });
    bus.unsubscribe(b);
    bus.subscribe([&](const std::string&) { log.push_back("D"); });
    bus.publish("x");
    EXPECT_EQ(log, (std::vector<std::string>{"A", "C", "D"}));
}

TEST(EventBus, SubscribeInsideACallbackCountsFromTheNextPublish) {
    EventBus bus;
    std::vector<std::string> log;
    bool added = false;
    bus.subscribe([&](const std::string&) {
        log.push_back("A");
        if (!added) {
            added = true;
            bus.subscribe([&](const std::string&) { log.push_back("B"); });
        }
    });

    EXPECT_EQ(bus.publish("first"), 1u) << "B joined mid-publish: not called by it";
    EXPECT_EQ(log, (std::vector<std::string>{"A"}));
    log.clear();
    EXPECT_EQ(bus.publish("second"), 2u);
    EXPECT_EQ(log, (std::vector<std::string>{"A", "B"}));
}

TEST(EventBus, UnsubscribeInsideACallbackCountsFromTheNextPublish) {
    EventBus bus;
    std::vector<std::string> log;
    long long b = 0;
    bool removed = false;
    bus.subscribe([&](const std::string&) {
        log.push_back("A");
        if (!removed) {
            removed = true;
            EXPECT_TRUE(bus.unsubscribe(b));
        }
    });
    b = bus.subscribe([&](const std::string&) { log.push_back("B"); });

    EXPECT_EQ(bus.publish("first"), 2u) << "B left mid-publish: still called by it";
    EXPECT_EQ(log, (std::vector<std::string>{"A", "B"}));
    log.clear();
    EXPECT_EQ(bus.publish("second"), 1u);
    EXPECT_EQ(log, (std::vector<std::string>{"A"}));
}

TEST(EventBus, CallbackCanUnsubscribeItself) {
    EventBus bus;
    int aCalls = 0;
    long long a = 0;
    a = bus.subscribe([&](const std::string&) {
        ++aCalls;
        bus.unsubscribe(a);
    });
    bus.subscribe([](const std::string&) {});

    EXPECT_EQ(bus.publish("first"), 2u);
    EXPECT_EQ(bus.publish("second"), 1u);
    EXPECT_EQ(aCalls, 1);
}

TEST(EventBus, CallbackCanPublish) {
    EventBus bus;
    std::vector<std::string> bLog;
    bus.subscribe([&](const std::string& e) {
        if (e == "outer") bus.publish("inner");
    });
    bus.subscribe([&](const std::string& e) { bLog.push_back(e); });

    EXPECT_EQ(bus.publish("outer"), 2u);
    // A handles "outer" by publishing "inner", which reaches B first.
    EXPECT_EQ(bLog, (std::vector<std::string>{"inner", "outer"}));
}

TEST(EventBus, OtherThreadsCanUseTheBusWhileACallbackRuns) {
    // The callback waits (up to 2 s) for another thread to subscribe. If
    // publish() holds a lock while calling back, that thread is stuck.
    EventBus bus;
    std::string outcome;
    std::thread helper;
    bus.subscribe([&](const std::string&) {
        std::promise<void> done;
        std::future<void> doneF = done.get_future();
        helper = std::thread([&bus, p = std::move(done)]() mutable {
            bus.subscribe([](const std::string&) {});
            p.set_value();
        });
        outcome = doneF.wait_for(2s) == std::future_status::ready ? "ok" : "blocked";
    });

    bus.publish("x");
    if (helper.joinable()) helper.join();
    EXPECT_EQ(outcome, "ok") << "another thread could not subscribe while a callback ran";
    EXPECT_EQ(bus.subscriberCount(), 2u);
}

TEST(EventBus, ConcurrentSubscribesGetUniqueIds) {
    constexpr int kThreads = 8;
    constexpr int kPerThread = 1000;
    EventBus bus;
    std::vector<std::vector<long long>> ids(kThreads);
    runTogether(kThreads, [&](int t) {
        for (int i = 0; i < kPerThread; ++i) {
            ids[t].push_back(bus.subscribe([](const std::string&) {}));
        }
    });

    std::set<long long> all;
    for (const auto& v : ids) all.insert(v.begin(), v.end());
    EXPECT_EQ(all.size(), static_cast<std::size_t>(kThreads * kPerThread));
    EXPECT_EQ(bus.subscriberCount(), static_cast<std::size_t>(kThreads * kPerThread));
}

TEST(EventBus, ConcurrentPublishAndChurn) {
    // 3 subscribers stay for the whole test and must hear every event
    // exactly once, while other threads keep subscribing and
    // unsubscribing around them.
    constexpr int kPublishers = 4;
    constexpr int kChurners = 4;
    constexpr int kPerThread = 2000;
    EventBus bus;
    std::atomic<int> heard[3] = {{0}, {0}, {0}};
    for (auto& h : heard) {
        bus.subscribe([&h](const std::string&) { h.fetch_add(1); });
    }
    std::atomic<bool> badCount{false};
    std::atomic<bool> badUnsubscribe{false};

    runTogether(kPublishers + kChurners, [&](int t) {
        for (int i = 0; i < kPerThread; ++i) {
            if (t < kPublishers) {
                if (bus.publish("tick") < 3) badCount = true;
            } else {
                long long id = bus.subscribe([](const std::string&) {});
                if (!bus.unsubscribe(id)) badUnsubscribe = true;
            }
        }
    });

    for (auto& h : heard) EXPECT_EQ(h.load(), kPublishers * kPerThread);
    EXPECT_FALSE(badCount) << "a publish missed a permanent subscriber";
    EXPECT_FALSE(badUnsubscribe) << "unsubscribe of a fresh id returned false";
    EXPECT_EQ(bus.subscriberCount(), 3u);
}

File: playground/java/days/day03/src/test/java/day03/EventBusTest.java
package day03;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.Future;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

import static day03.RunTogether.inBackground;
import static day03.RunTogether.runTogether;
import static java.util.concurrent.TimeUnit.SECONDS;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Timeout(value = 60, threadMode = Timeout.ThreadMode.SEPARATE_THREAD)
class EventBusTest {

    @Test
    void publishCallsSubscribersInOrder() {
        EventBus bus = new EventBus();
        List<String> log = new ArrayList<>();
        bus.subscribe(e -> log.add("A:" + e));
        bus.subscribe(e -> log.add("B:" + e));
        bus.subscribe(e -> log.add("C:" + e));
        assertEquals(3, bus.publish("x"));
        assertEquals(List.of("A:x", "B:x", "C:x"), log);
    }

    @Test
    void publishWithNoSubscribersCallsNobody() {
        EventBus bus = new EventBus();
        assertEquals(0, bus.publish("x"));
        assertEquals(0, bus.subscriberCount());
    }

    @Test
    void idsArePositiveAndUnique() {
        EventBus bus = new EventBus();
        Set<Long> ids = new HashSet<>();
        for (int i = 0; i < 100; i++) {
            long id = bus.subscribe(e -> {});
            assertTrue(id > 0, "ids must be > 0, got " + id);
            ids.add(id);
        }
        assertEquals(100, ids.size(), "different ids among 100 subscribes");
        assertEquals(100, bus.subscriberCount());
    }

    @Test
    void unsubscribeStopsDelivery() {
        EventBus bus = new EventBus();
        List<String> log = new ArrayList<>();
        bus.subscribe(e -> log.add("A"));
        long b = bus.subscribe(e -> log.add("B"));
        bus.subscribe(e -> log.add("C"));

        assertTrue(bus.unsubscribe(b));
        assertFalse(bus.unsubscribe(b), "already removed");
        assertFalse(bus.unsubscribe(12345), "never existed");
        assertEquals(2, bus.subscriberCount());
        assertEquals(2, bus.publish("x"));
        assertEquals(List.of("A", "C"), log);
    }

    @Test
    void orderIsKeptAfterUnsubscribeAndSubscribe() {
        EventBus bus = new EventBus();
        List<String> log = new ArrayList<>();
        bus.subscribe(e -> log.add("A"));
        long b = bus.subscribe(e -> log.add("B"));
        bus.subscribe(e -> log.add("C"));
        bus.unsubscribe(b);
        bus.subscribe(e -> log.add("D"));
        bus.publish("x");
        assertEquals(List.of("A", "C", "D"), log);
    }

    @Test
    void subscribeInsideACallbackCountsFromTheNextPublish() {
        EventBus bus = new EventBus();
        List<String> log = new ArrayList<>();
        AtomicBoolean added = new AtomicBoolean(false);
        bus.subscribe(e -> {
            log.add("A");
            if (!added.getAndSet(true)) {
                bus.subscribe(e2 -> log.add("B"));
            }
        });

        assertEquals(1, bus.publish("first"), "B joined mid-publish: not called by it");
        assertEquals(List.of("A"), log);
        log.clear();
        assertEquals(2, bus.publish("second"));
        assertEquals(List.of("A", "B"), log);
    }

    @Test
    void unsubscribeInsideACallbackCountsFromTheNextPublish() {
        EventBus bus = new EventBus();
        List<String> log = new ArrayList<>();
        AtomicLong b = new AtomicLong();
        AtomicBoolean removed = new AtomicBoolean(false);
        AtomicBoolean unsubscribeReturned = new AtomicBoolean(false);
        bus.subscribe(e -> {
            log.add("A");
            if (!removed.getAndSet(true)) {
                unsubscribeReturned.set(bus.unsubscribe(b.get()));
            }
        });
        b.set(bus.subscribe(e -> log.add("B")));

        assertEquals(2, bus.publish("first"), "B left mid-publish: still called by it");
        assertEquals(List.of("A", "B"), log);
        assertTrue(unsubscribeReturned.get(), "unsubscribe(b) inside a callback returned false");
        log.clear();
        assertEquals(1, bus.publish("second"));
        assertEquals(List.of("A"), log);
    }

    @Test
    void callbackCanUnsubscribeItself() {
        EventBus bus = new EventBus();
        AtomicInteger aCalls = new AtomicInteger();
        AtomicLong a = new AtomicLong();
        a.set(bus.subscribe(e -> {
            aCalls.incrementAndGet();
            bus.unsubscribe(a.get());
        }));
        bus.subscribe(e -> {});

        assertEquals(2, bus.publish("first"));
        assertEquals(1, bus.publish("second"));
        assertEquals(1, aCalls.get());
    }

    @Test
    void callbackCanPublish() {
        EventBus bus = new EventBus();
        List<String> bLog = new ArrayList<>();
        bus.subscribe(e -> {
            if (e.equals("outer")) bus.publish("inner");
        });
        bus.subscribe(bLog::add);

        assertEquals(2, bus.publish("outer"));
        // A handles "outer" by publishing "inner", which reaches B first.
        assertEquals(List.of("inner", "outer"), bLog);
    }

    @Test
    void otherThreadsCanUseTheBusWhileACallbackRuns() throws Exception {
        // The callback waits (up to 2 s) for another thread to subscribe. If
        // publish() holds a lock while calling back, that thread is stuck.
        EventBus bus = new EventBus();
        AtomicReference<String> outcome = new AtomicReference<>();
        AtomicReference<Future<Long>> helper = new AtomicReference<>();
        bus.subscribe(e -> {
            Future<Long> h = inBackground(() -> bus.subscribe(e2 -> {}));
            helper.set(h);
            try {
                h.get(2, SECONDS);
                outcome.set("ok");
            } catch (TimeoutException stuck) {
                outcome.set("blocked");
            } catch (Exception other) {
                outcome.set("helper failed: " + other);
            }
        });

        bus.publish("x");
        assertNotNull(helper.get(), "publish() did not call the callback");
        helper.get().get(); // wait for the helper thread to finish
        assertEquals("ok", outcome.get(), "another thread could not subscribe while a callback ran");
        assertEquals(2, bus.subscriberCount());
    }

    @Test
    void concurrentSubscribesGetUniqueIds() throws Exception {
        int threads = 8;
        int perThread = 1000;
        EventBus bus = new EventBus();
        long[][] ids = new long[threads][perThread];
        runTogether(threads, t -> {
            for (int i = 0; i < perThread; i++) {
                ids[t][i] = bus.subscribe(e -> {});
            }
        });

        Set<Long> all = new HashSet<>();
        for (long[] mine : ids) {
            for (long id : mine) all.add(id);
        }
        assertEquals(threads * perThread, all.size(), "different ids handed out");
        assertEquals(threads * perThread, bus.subscriberCount());
    }

    @Test
    void concurrentPublishAndChurn() throws Exception {
        // 3 subscribers stay for the whole test and must hear every event
        // exactly once, while other threads keep subscribing and
        // unsubscribing around them.
        int publishers = 4;
        int churners = 4;
        int perThread = 2000;
        EventBus bus = new EventBus();
        AtomicInteger[] heard = {new AtomicInteger(), new AtomicInteger(), new AtomicInteger()};
        for (AtomicInteger h : heard) {
            bus.subscribe(e -> h.incrementAndGet());
        }
        AtomicBoolean badCount = new AtomicBoolean(false);
        AtomicBoolean badUnsubscribe = new AtomicBoolean(false);

        runTogether(publishers + churners, t -> {
            for (int i = 0; i < perThread; i++) {
                if (t < publishers) {
                    if (bus.publish("tick") < 3) badCount.set(true);
                } else {
                    long id = bus.subscribe(e -> {});
                    if (!bus.unsubscribe(id)) badUnsubscribe.set(true);
                }
            }
        });

        for (AtomicInteger h : heard) {
            assertEquals(publishers * perThread, h.get(), "events heard by a permanent subscriber");
        }
        assertFalse(badCount.get(), "a publish missed a permanent subscriber");
        assertFalse(badUnsubscribe.get(), "unsubscribe of a fresh id returned false");
        assertEquals(3, bus.subscriberCount());
    }
}
